db/map.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (tx WrappedTransaction) Rollback() error {
+	return tx.Transaction.Rollback()
+}
+
+func (tx WrappedTransaction) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Get(holder, keys...)
+}
+
+func (tx WrappedTransaction) Insert(list ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Insert(list...)
+}
+
+func (tx WrappedTransaction) Update(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Update(list...)
+}
+
+func (tx WrappedTransaction) Delete(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Delete(list...)
+}
+
+func (tx WrappedTransaction) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Select(holder, query, args...)
+}
+
+func (tx WrappedTransaction) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).SelectOne(holder, query, args...)
+}
+
+func (tx WrappedTransaction) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Exec(query, args...)
+}
+
+var (
+	// selectTableRegexp matches the table name from an SQL select statement
+	selectTableRegexp = regexp.MustCompile(`(?i)^\s*select\s+[a-z\d\.\(\), \_\*` + "`" + `]+\s+from\s+([a-z\d\_,` + "`" + `]+)`)
+	// insertTableRegexp matches the table name from an SQL insert statement
+	insertTableRegexp = regexp.MustCompile(`(?i)^\s*insert\s+into\s+([a-z\d \_,` + "`" + `]+)\s+(?:set|\()`)
+	// updateTableRegexp matches the table name from an SQL update statement
+	updateTableRegexp = regexp.MustCompile(`(?i)^\s*update\s+([a-z\d \_,` + "`" + `]+)\s+set`)
+	// deleteTableRegexp matches the table name from an SQL delete statement
+	deleteTableRegexp = regexp.MustCompile(`(?i)^\s*delete\s+from\s+([a-z\d \_,` + "`" + `]+)\s+where`)
+
+	// tableRegexps is a list of regexps that tableFromQuery will try to use in
+	// succession to find the table name for an SQL query.
+	tableRegexps = []*regexp.Regexp{
+		selectTableRegexp,
+		insertTableRegexp,
+		updateTableRegexp,
+		deleteTableRegexp,
+	}
+)
+
+// tableFromQuery uses the tableRegexps on the provided query to return the
+// associated table name or an empty string if it can't be determined from the
+// query.
+func tableFromQuery(query string) string {
+	for _, r := range tableRegexps {
+		if matches := r.FindStringSubmatch(query); len(matches) >= 2 {
+			return matches[1]
+		}
+	}
+	return ""
+}
+
+// WrappedExecutor wraps a gorp.SqlExecutor such that its major functions
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedExecutor struct {
+	gorp.SqlExecutor
+}
+
+func (we WrappedExecutor) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
```

Comment:
It looks like Get, Update, Insert, and Delete don't call `tableFromQuery`. It seems like maybe that's because we usually have a type we can use instead of going through the regexps, but we do have regexps defined for insert and delete. Did you intend to add tableFromQuery calls to those functions?

Reply:
> Did you intend to add tableFromQuery calls to those functions?

No, but I wanted to catch the cases where someone `Exec`'s an insert or delete. Those do require going through `tableFromQuery`.

Reply:
Thanks for the explanation. I think this makes sense to explain in a comment somewhere.